File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> freq = new HashMap<Integer,Integer>();
        for(int i =0 ; i < nums.length ; i++){
            if(freq.keySet().contains(nums[i])){
                freq.put( nums[i], freq.get(nums[i]) + 1);
            }
            else{
                freq.put( nums[i], 1);
            }
        }
        return getKeys(freq,k);
    }
    
    public int[] getKeys(Map<Integer,Integer> freq, int k){
        List<Integer> temp = new ArrayList<Integer>(); 
        for(int m =1; m <= k ;m++){
            int val = (Collections.max(freq.entrySet(),(entry1,entry2) -> entry1.getValue() - entry2.getValue()).getKey());
            temp.add(val);
            freq.remove(val);
        }
        
        int[] keys = new int[temp.size()];
        int j =0;
        for(int i : temp){
           keys[j] = i;
           j = j + 1;
        }
        return keys;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
from collections import Counter
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # counter=Counter(nums)
        counter={}
        for i in range(len(nums)):
            if nums[i] not in counter:
                counter[nums[i]]=1
            else:
                counter[nums[i]]+=1
        min_heap=[]
        for key,value in counter.items():
            if len(min_heap)<k:
                heapq.heappush(min_heap,(value,key))
            else:
                heapq.heappushpop(min_heap,(value,key))
        return [h[1] for h in min_heap]
        #tc:N+Nlogk--->Nlogk assuming k<N
        #sc:k+N-->N
        